modules/magnolia/src/main/scala/pureconfig/module/magnolia/EnumerationConfigReaderBuilder.scala

```diff
@@ -0,0 +1,49 @@
+package pureconfig.module.magnolia
+
+import scala.language.experimental.macros
+
+import magnolia._
+import pureconfig.ConfigReader.Result
+import pureconfig.generic.error.NoValidCoproductOptionFound
+import pureconfig.{ConfigCursor, ConfigReader}
+
+/**
+  * A type class to build `ConfigReader`s for sealed families of case objects where each type is encoded as a
+  * `ConfigString` based on the type name.
+  *
+  * @tparam A the type of objects capable of being read as an enumeration
+  */
+private[magnolia] trait EnumerationConfigReaderBuilder[A] {
+  def build(transformName: String => String): ConfigReader[A]
+}
+
+object EnumerationConfigReaderBuilder {
+  type Typeclass[A] = EnumerationConfigReaderBuilder[A]
+
+  def combine[A](ctx: CaseClass[EnumerationConfigReaderBuilder, A]): EnumerationConfigReaderBuilder[A] =
+    new EnumerationConfigReaderBuilder[A] {
+      def build(transformName: String => String): ConfigReader[A] =
+        new ConfigReader[A] {
+          def from(cur: ConfigCursor): Result[A] =
+            if (ctx.isObject) Right(ctx.rawConstruct(Seq.empty))
+            else cur.asConfigValue.right.flatMap(v => cur.failed(NoValidCoproductOptionFound(v, Seq.empty)))
```

Comment:
How are you planning to handle valid coproduct options in errors? I know you mentioned revisiting this after #787; isn't it more convenient to do this now that we're adding a new implementation? Or do you prefer to do it on a separate PR?

Reply:
Also, is `NoValidCoproductOptionFound` the right error here?

If I understood the code correctly this is the case where we find ourselves deriving an enum reader for a case class rather than an object. In this case, shouldn't we throw an exception to cause the derivation to fail at compile-time?

Reply:
> Also, is `NoValidCoproductOptionFound` the right error here?
> 
> If I understood the code correctly this is the case where we find ourselves deriving an enum reader for a case class rather than an object. In this case, shouldn't we throw an exception to cause the derivation to fail at compile-time?

Unfortunately, an exception thrown here will only be visible at run time. I don't think there's a straightforward way to fail the compilation at this point. I suppose we can create a custom `FailureReason` for this, but ideally we wouldn't be deriving type classes for case classes.

Reply:
That's unexpected :/ I think we really need this error to stand out, as the behavior differs from shapeless and from the compile-time safety of PureConfig in general. I'd probably go for a different error where we can specify a reason, like `CannotConvert`. However, if you still feel we're ok `NoValidCoproductOptionFound` I'm ok with it too.

It's probably worth asking in magnolia how to make this use case possible, too.

Reply:
I opened an issue in Magnolia (propensive/magnolia#267) to see if this use case would be possible. In propensive/magnolia#263, they introduced the possibility to only define the `dispatch` method. I suppose we could have an implementation for `dispatch` only and provide an alternative derivation method for objects. However, I think we'd still need a macro to provide an evidence that a given type is an object.

For the time being, I've changed the failure reason to a `CannotConvert`, so that we can make the underlying issue for the error more explicit.